### DISCUSSION TITLE
Advanced usage: keep Twig templates compiled/static in `var/build` directory to speed up local environment

### DISCUSSION THREAD

**Discussion by ruudk** [symfony/symfony#50629] [Q&A] [answered]
We have a large application (±11 years, 12.000 classes, Symfony 6.3, PHP 8.2) and experience slowdowns of cache compilation. While doing some profiling I noticed that 23% of the time it takes to generate the cache is coming from `Symfony\Bundle\TwigBundle\CacheWarmer\TemplateCacheWarmer`. That's quite a lot! 

The thing is, our Twig templates rarely change. 😅 

If we could shave off 20% of our cache compilation time by having a fixed set of compiled Twig templates (the raw PHP files) that would be a huge win. 

We can use a file watcher to detect changes to `.twig` files to make sure the cache is re-generated.

Before I do the work, I wonder: Is there anybody that ever tried this? Or are there already ways to allow this?

**Comment by ghost** [comment]
I have the same problem in my project (sf 3.4, php 7.4, 10k+ php files, lots of templates). The `cache:clear` takes a lot of time and Twig's warmer is the heavies one. However I've never had time to search for possible solutions.

**Comment by MatTheCat** [comment]
If the debug mode is enabled, the [`auto_reload`](https://symfony.com/doc/current/reference/configuration/twig.html#auto-reload) option should kick in, which mean you could disable the twig cache warmer through a compiler pass 🤔

I never did this and cannot test it right now but it sounds legit in my head.

**Reply by MatTheCat** [reply, marked as answer]
Just confirmed it works. Adding

```php
if ($this->debug && $container->hasDefinition('twig.template_cache_warmer')) {
    $container->removeDefinition('twig.template_cache_warmer');
}
```

in a compiler pass will prevent twig cache to be warmed in debug mode. `auto_reload` will then ensure the templates’ cache is generated/refreshed if needed.

**Reply by ruudk** [reply]
Thank you, this works!

**Reply by MatTheCat** [reply]
👌
Also note the `cache:clear` command exposes a `no-optional-warmers` option. This would disable twigʼs cache warmup, amongst others.

**Comment by 94noni** [comment]
with a config like this in the twig, to avoid recompiling on dev env
https://github.com/symfony/symfony/pull/43138
https://symfony.com/blog/new-in-symfony-5-4-profiler-improvements#conditional-profiling
indeed sound doable but never tried it neither

**Reply by ghost** [reply]
How disabling the web profiler has an effect on recompiling twig templates? They have to be recompiled every time they change.

**Comment by 94noni** [comment]
nono I mean with same config logic, something to toggle on/off to let the framework rebuild twig cache or not
as the profiler config
